lib/src/rules/voidness_propagation.dart

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2018, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Do not loose voidness.';
+
+const _details = r'''
+
+**DO** not loose voidness. This is likely to lead to a runtime error.
+
+**BAD:**
+
+```dart
+List<void> x;
+List<int> y = x;
+```
+
+```dart
+List<void> x;
+List<Object> y = x;
+```
+
+''';
+
+class VoidnessPropagation extends LintRule implements NodeLintRule {
+  VoidnessPropagation()
+      : super(
+            name: 'voidness_propagation',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry) {
+    final visitor = new _Visitor(this);
+    registry.addAssignmentExpression(this, visitor);
+    registry.addExpressionFunctionBody(this, visitor);
+    registry.addInstanceCreationExpression(this, visitor);
+    registry.addMethodDeclaration(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+    registry.addReturnStatement(this, visitor);
+    registry.addVariableDeclaration(this, visitor);
```

Comment:
I would like to maybe see about giving this lint better support in the analyzer API.

There are many more places that downcasts can occur -- for instance:

```
class DefinesPlus {
  operator +(DefinesPlus other) => null;
}

void main() {
  var x = DefinesPlus();
  void y;
  print(x + y); // downcast from void to DefinesPlus
}
```

The CFE project that lowers Dart to kernel code will include an explicit cast here, and the analyzer can do the same sort of resolution. If we could expose that in the tree, that would make it easier for you to, for instance, find any expressions that are downcasted and then check what the from + to types were. This would get a performance boost (by not having to check every node for downcasts _again_), make this code simpler (hopefully!), and make it easy to have this lint and others like it be consistently conclusive in coverage.

Are you interested in doing that for this lint or do you want me to keep that in my mind as a theoretical? Note: I have not discussed this with the other analyzer teammembers and they may not agree with me on it.

Reply:
> I would like to maybe see about giving this lint better support in the analyzer API.

I'm not sure what kind of support you're thinking about, so it's hard for me to say what I think about it. :-)

I would expect that we already have the "from" and "to" types in the AST. It should just be a matter of testing the relationships between those types. You might be proposing that we encapsulate that test in a `requiresImplicitDowncast` member so that clients don't have to repeat the test in multiple locations and risk getting it wrong, which seems reasonable.